MaskedInput

amottola opened the pull request:
Following discussion in #4591, I finally found some time to refactor the original proposal and introduce here the `MaskedInput` widget, derived from `Input`. `MaskedInput` behaves like a templated `Input` but does not support `highlighter`, `password`, `restrict`, `type`, `max_length` and `suggester` parameters, greatly simplifying logic and making the control more distinct from standard `Input`.

You will also note I updated docs accordingly, also adding the widget to the gallery. A word about docs and licensing: I rewrote the docs so there is nothing left of the original version; this should hopefully settle the issue reported by @merriam. 

**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [x] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

willmcgugan reviewed (changes requested):
Sorry for the delay. This is amazing work, and a really useful feature.

I've left some requests.

Could you also confirm if this works with double width characters? i.e. Asian characters and emoji. It is a common source of issues for this kind of thing.

amottola commented:
Hi. I've hopefully addresses the review issues, let me know if it's ok.
I did not check for double width characters as I have zero experience with them... any help is appreciated, otherwise I'll have to find some time to investigate this later.

I've had problems rebasing: now my fork is 66 commits ahead and 74 behind, and github tells me I cannot sync because of conflicts... I did a sync of my fork `main` branch and rebased my `masked_input` branch on it, so I do not have any more conflicts there, but this did not help as github still tells me I cannot sync... my git-fu is not so strong after all 😞

willmcgugan commented:
@amottola Sorry for the slow response, again! There are a few test files. Would you mind looking at those?

Once the tests are passing, I think this should be good to go.

amottola commented:
@willmcgugan No need to apologize, really; thanks for your time indeed. I just fixed the tests, should be all ready to go now.

darrenburns commented:
I've resolved the changelog merge conflict - @willmcgugan are we good to go?

TomJGooding commented on docs/widgets/masked_input.md:
```diff
@@ -0,0 +1,84 @@
+# MaskedInput
+
+!!! tip "Added in version 0.74.0"
```
Just to note that this version will need updating before release

darrenburns replied:
Cheers @TomJGooding - I've updated that now.

willmcgugan commented:
@amottola Thanks so much. This is an awesome feature!
